### PR TITLE
WIP: Item Stats Rework

### PR DESCRIPTION
**Pull request by Zintixx**
These changes are mainly to incorporate the basic attributes (divided between static options and constant options).

Rarity is no longer grabbed from the constant ID - that's for constant options. The closest thing I can find to a indication of what the "default" rarity is, is from itemwebfinder, so this is where I'm grabbing it from.

STILL WIP:
- This is currently still missing some default basic attributes for items. Still trying to figure out how it's calculated from outside of itemoptions xmls.
- Constant and Random options have a range of stats that needs to be implemented. I'm currently just grabbing the lowest value for the sake of trying to get everything right the first time around.
- Fix `ChangeAttributesHandler.HandleChangeAttributues` I broke this switching out of structs to classes. I'm not entirely sure how to go about this now that it's a class.
- Still need to apply some stat variation to basic attributes

###Misc
- Some progress on Gemstones have begun. I mainly paused on this to get ItemStats reworked
- Added FieldpopupEvent as an event. This is how the event map UI icon appears to players. Currently it wont take you anywhere - it's not fully implemented.

### CODE REVIEW

File: GameDataParser/Parsers/ItemOptionRandomParser.cs [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+                            case "hiddenbapadd_value_base":
+                            case "hiddennddadd_value_base":
+                            case "hiddenwapadd_value_base":
+                            case "nddcalibrationfactor_rate_base":
+                            case "wapcalibrationfactor_rate_base":
+                            case "conditionreduce_rate_base":
+                            case "sgi_target":
+                                break;
+                            default:
+                                System.Console.WriteLine(item.Name);
```
Comment (AngeloTadeucci) [inline GameDataParser/Parsers/ItemOptionRandomParser.cs:596]: Intended?

File: MapleServer2/Database/DatabaseContext.cs [hunk 2]
```diff
@@ -348,8 +349,8 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
                    i => i == null ? new MusicScore() : JsonConvert.DeserializeObject<MusicScore>(i));
 
                 entity.Property(e => e.Stats).HasConversion(
-                    i => JsonConvert.SerializeObject(i, settings),
-                    i => i == null ? new ItemStats() : JsonConvert.DeserializeObject<ItemStats>(i, settings));
```
Comment (AngeloTadeucci) [inline MapleServer2/Database/DatabaseContext.cs:352]: If the settings are not needed anymore, you can remove line 57.

File: MapleServer2/Packets/Helpers/ItemPacketHelper.cs [hunk 3]
```diff
@@ -135,17 +135,22 @@ private static PacketWriter WriteAppearance(this PacketWriter pWriter, Item item
...
-            foreach (SpecialStat stat in specialBasicStats)
+
+            List<ItemStat> basicSpecialStats = stats.BasicStats.Where(x => x.GetType() == typeof(SpecialStat)).ToList();
+            pWriter.WriteShort((short) basicSpecialStats.Count);
+            foreach (SpecialStat stat in basicSpecialStats)
             {
-                pWriter.Write(stat);
+                pWriter.WriteShort((short) stat.ItemAttribute);
+                pWriter.WriteFloat(stat.Percent);
+                pWriter.WriteInt(stat.Flat);
```
Comment (AngeloTadeucci) [inline MapleServer2/Packets/Helpers/ItemPacketHelper.cs:153]: SpecialStat was wrong all this time? It isn't two floats?

File: GameDataParser/Parsers/ItemGenstoneUpgradeParser.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                        metadata.IngredientAmounts.Add(int.Parse(key.Attributes["IngredientCount2"].Value));
+                    }
+
+                    if (key.Attributes["IngredientItemID3"] != null)
+                    {
+                        metadata.IngredientItems.Add(Regex.Match(key.Attributes["IngredientItemID3"].Value, @"[a-zA-Z]+").Value);
+                        metadata.IngredientAmounts.Add(int.Parse(key.Attributes["IngredientCount3"].Value));
+                    }
+
+                    if (key.Attributes["IngredientItemID4"] != null)
```
Comment (AlanMorel) [inline GameDataParser/Parsers/ItemGenstoneUpgradeParser.cs:57]: Could be turned into a for loop, 1-4

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Packets/Helpers/ItemPacketHelper.cs (hunk 3 above)
Comment (Zintixx) [reply MapleServer2/Packets/Helpers/ItemPacketHelper.cs:153]: I just realized that it is two floats... but looking in game, it seems to accept it and display the value correctly for flats.

File: MapleServer2/Database/DatabaseContext.cs (hunk 2 above)
Comment (Zintixx) [reply MapleServer2/Database/DatabaseContext.cs:352]: i'm dumb nvm it's still needed

File: MapleServer2/Packets/Helpers/ItemPacketHelper.cs (hunk 3 above)
Comment (Zintixx) [reply MapleServer2/Packets/Helpers/ItemPacketHelper.cs:153]: I suppose it's the correct thing to do to change it to floats.
sadge

**Comment by AngeloTadeucci** [comment]
Just ran GameDataParser and it generated an new item hash.

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**